osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs

```diff
@@ -146,34 +133,41 @@ private void applyRandomisation(float rateOfChangeMultiplier, RandomObjectInfo p
-            var minMargin = new MarginPadding();
+            var minMargin = getMinSliderMargin(slider);
 
-            foreach (var hitObject in slider.NestedHitObjects.Where(o => o is SliderTick || o is SliderEndCircle))
-            {
-                if (!(hitObject is OsuHitObject osuHitObject))
-                    continue;
+            slider.Position = new Vector2(
+                Math.Clamp(slider.Position.X, minMargin.Left, OsuPlayfield.BASE_SIZE.X - minMargin.Right),
+                Math.Clamp(slider.Position.Y, minMargin.Top, OsuPlayfield.BASE_SIZE.Y - minMargin.Bottom)
+            );
 
-                var relativePos = Vector2.Subtract(osuHitObject.Position, slider.Position);
+            currentObjectInfo.PositionRandomised = slider.Position;
+            currentObjectInfo.EndPositionRandomised = slider.EndPosition;
 
-                minMargin.Left = Math.Max(minMargin.Left, -relativePos.X);
-                minMargin.Right = Math.Max(minMargin.Right, relativePos.X);
-                minMargin.Top = Math.Max(minMargin.Top, -relativePos.Y);
-                minMargin.Bottom = Math.Max(minMargin.Bottom, relativePos.Y);
-            }
+            shiftNestedObjects(slider, currentObjectInfo.PositionRandomised - currentObjectInfo.PositionOriginal);
+        }
+
+        /// <summary>
+        /// Calculates the min. distances from the <see cref="Slider"/>'s position to the playfield border for the slider to be fully inside of the playfield.
+        /// </summary>
+        private MarginPadding getMinSliderMargin(Slider slider)
+        {
+            var pathPositions = new List<Vector2>();
+            slider.Path.GetPathToProgress(pathPositions, 0, 1);
 
-            if (slider.Position.X < minMargin.Left)
-                slider.Position = new Vector2(minMargin.Left, slider.Position.Y);
-            else if (slider.Position.X + minMargin.Right > OsuPlayfield.BASE_SIZE.X)
-                slider.Position = new Vector2(OsuPlayfield.BASE_SIZE.X - minMargin.Right, slider.Position.Y);
+            var minMargin = new MarginPadding();
+
+            foreach (var pos in pathPositions)
+            {
+                minMargin.Left = Math.Max(minMargin.Left, -pos.X);
+                minMargin.Right = Math.Max(minMargin.Right, pos.X);
+                minMargin.Top = Math.Max(minMargin.Top, -pos.Y);
+                minMargin.Bottom = Math.Max(minMargin.Bottom, pos.Y);
+            }
 
-            if (slider.Position.Y < minMargin.Top)
-                slider.Position = new Vector2(slider.Position.X, minMargin.Top);
-            else if (slider.Position.Y + minMargin.Bottom > OsuPlayfield.BASE_SIZE.Y)
-                slider.Position = new Vector2(slider.Position.X, OsuPlayfield.BASE_SIZE.Y - minMargin.Bottom);
+            minMargin.Left = Math.Min(minMargin.Left, OsuPlayfield.BASE_SIZE.X - minMargin.Right);
+            minMargin.Top = Math.Min(minMargin.Top, OsuPlayfield.BASE_SIZE.Y - minMargin.Bottom);
```

Comment:
This whole thing reads.... really badly.

I don't even know where to start but there has to be a better way of writing this.

- Why are you using `-pos` here? Just because it happens to be subtracted at a later point?
- What are the final subtractions for? And why are they done here while the other two are done in the other method above?

The original flow here I could actually follow. This I can't.

Reply:
1. `-pos` is used because if e.g. `slider.Position.X` is positive, `minMargin.Left` **shouldn't change**, because (since `pos` is realtive to `slider.Position`) `pos` is to the right of `slider.Position`. `minMargin.Left` is positive, `-pos.X` is negative (and thus smaller than `minMargin.Left`), so `Math.Max(minMargin.Left, -pos.X)` equals `minMargin.Left` and it **stays the same**. If `slider.Position.X` is *negative* however, the already negative `pos.X` becomes positive, which means `minMargin.Left` is positive as well (which it should be because a negative margin doesn't really make sense).
2. https://github.com/ppy/osu/blob/b214f2ae0e39a08e6763c8943d9e63991b996568/osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs#L167-L168 These are to make sure that e.g. `minMargin.Left + minMargin.Right` isn't bigger than `OsuPlayfield.BASE_SIZE.X`, because if they were, `Math.Clamp()` would throw a `System.ArgumentException` because `min` would be bigger than `max`. (I actually just noticed I could use `MathHelper` instead, which doesn't throw an Exception.) https://github.com/ppy/osu/blob/b214f2ae0e39a08e6763c8943d9e63991b996568/osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs#L139-L140 I put it into `getMinSliderMargin()` because `getMinSliderMargin()` should just return the final margin so that `moveSliderIntoPlayfield()` doesn't have to deal with any margin calculations.

I actually thought this would be easier to follow than the original code, because the calculation of the margins essentially stayed the same (the "`-pos` thing" you mentiond was already there before), I basically just moved it into its own method and I'm using `Math.Clamp()` now because it's way shorter than this:
https://github.com/ppy/osu/blob/6ebac0b4628162e90f5b06e2a63e8a276416d3ec/osu.Game.Rulesets.Osu/Mods/OsuModRandom.cs#L165-L173
(And of course I also changed the positions that are used for calculating the margins, which is the main point of this PR.)

Reply:
Explaining here is fine and all, but code should be stand-alone readable.

It's a pretty isolated case so whatever, but I'm sure there's a better way it could be written. I guess if others are okay with the way this is structured then it's fine from a functionality perspective.

Reply:
I also find this new logic confusing, for what it's worth. From what I can tell, the for loop is basically computing the bounding box of the path, but with the `Left` and `Top` coordinate signs flipped. I think if the variables were actually called what they are, as in:

```csharp
var boundingBox = GetSurroundingQuad(pathPositions);

var minMargin = new MarginPadding
{
    Left = Math.Min(-pathBoundingBox.Left, OsuPlayfield.BASE_SIZE.X - pathBoundingBox.Right),
    Top = Math.Min(-pathBoundingBox.Top, OsuPlayfield.BASE_SIZE.Y - pathBoundingBox.Bottom)
};
```

where `GetSurroundingQuad()` is something like what `SelectionHandler` [already has](https://github.com/ppy/osu/blob/master/osu.Game/Screens/Edit/Compose/Components/SelectionHandler.cs#L422-L444), this would be way more understandable.